### ISSUE TITLE
Incremental parsing creates error node when fresh parse doesn't

### ISSUE THREAD

**Issue by casouri** [tree-sitter/tree-sitter#4001]
### Problem

I'm not sure if this is a tree-sitter-haskell bug or tree-sitter bug. I guess in theory you can write the grammar in an ambiguous way such that it generates wrong parse tree on edits. But Haskell grammar dev wasn't responding (https://github.com/tree-sitter/tree-sitter-haskell/issues/129), so I'm posting it here so the issue gets more visibility.

### Steps to reproduce

Here's what I did:

At the beginning, there's this file:

```haskll
f x =
              let a = 2
                  c = 1
 in a
```

I create a parser and parse it, got a parse tree.

Then I edit the file, indent the second and third lines:

```haskell
f x =
  let a = 2
      c = 1
 in a
```

I parse the file again incrementally. Now the parse tree I get has an error node in it, even though the source code is syntactically correct.

More over, if I now create a fresh parser, and parse the edited file anew, the parse tree I get is different from the one I got from incremental parsing, and it doesn't have any error nodes.

I created a reproduce repo at here: https://github.com/casouri/tree-sitter-haskell-repo

Running run.sh should clone and compile tree-sitter and tree-sitter-haskell, and compile and run recipe.c. Running it should print out the two parse tree I get, the first one from incremental parsing, the second one from a fresh parse.

```
Parse tree of edited file:
(haskell declarations: (declarations (function name: (variable) patterns: (patterns (variable)) match: (match expression: (let_in binds: (local_binds decl: (bind name: (variable) match: (match expression: (literal (integer))))) (ERROR (variable)) expression: (variable))))))

Parse tree of a fresh parse:
(haskell declarations: (declarations (function name: (variable) patterns: (patterns (variable)) match: (match expression: (let_in binds: (local_binds decl: (bind name: (variable) match: (match expression: (literal (integer)))) decl: (bind name: (variable) match: (match expression: (literal (integer))))) expression: (variable))))))
```

### Expected behavior

The two parse tree should be the same.

### Tree-sitter version (tree-sitter --version)

latest

### Operating system/version

macOS 14.6.1 (23G93)

**Comment by maxbrunsfeld** [comment]
I think this is indeed a Tree-sitter bug. Since it is manifesting with Tree-sitter-haskell, we're probably incorrectly re-using a node that is dependent on the column position of an external token. There have been bugs like this in the past. AFAIK, the most recent attempt to correct the behavior is https://github.com/tree-sitter/tree-sitter/pull/3257.

**Comment by amaanq** [comment]
I think this isn't particularly a core library issue, but rather because you're using dummy points, `(0, 0)`, and only relying on the byte offset. Unfortunately, for languages where indentation (and subsequently, column position) matters, tree-sitter relies on the row and column information to determine if a node has changes or not (see [here](https://github.com/tree-sitter/tree-sitter/blob/29e6717c3150c6c8f2855c5ab679569c5654c8b8/lib/src/subtree.c#L682) and [here](https://github.com/tree-sitter/tree-sitter/blob/29e6717c3150c6c8f2855c5ab679569c5654c8b8/lib/src/subtree.c#L781)). I think you *really* should start passing in the correct point data for edits, as this issue works correctly if you do then. I know we try to make parsing and lexing functionality dependent only on byte offset information typically, but this is one case where we *do* need row/col information.

**Comment by casouri** [comment]
Ok, makes sense. I appreciate the explanation. Emacs uses gap buffer internally so getting row and column isn't easy, but I'll see what I can do.

**Comment by amaanq** [comment]
I see, I think in theory we *could* make this purely dependent on byte information, but we'd probably incur a performance hit as a result of that. Hopefully you can make it work in Emacs

**Comment by casouri** [comment]
Yeah, line/col info needs to be track by either the editor or tree-sitter, since a lot of editors already track that, I can see tree-sitter not wanting to track it internally.

**Comment by maxbrunsfeld** [comment]
Yup, Amaan is right. Mostly Tree-sitter does not rely on the point coordinates (except for providing accurate points on the nodes). But for languages where the external scanner checks the current column value, the node invalidation actually does depend on the points to at are passed during edits.

We should probably make that clearer in the docs; I hadn’t thought about how non-obvious it was until now.
